keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1763 @@
+  resource drivers and which parameters they support. Portability of
+  workloads could be added on top of this proposal by introducing the
+  selection of a resource implementation through labels and
+  standardizing those labels and the associated parameters. The
+  [Resource Class
+  Proposal](https://docs.google.com/document/d/1qKiIVs9AMh2Ua5thhtvWqOqW0MSle_RV3lfriO1Aj6U/edit#heading=h.jzfmfdca34kj)
+  included such an approach.
+
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal is that a resource driver handles all operations that are specific
+to the resources managed by that driver. This includes operations at
+the control plane level (tracking where in the cluster resources are
+available, helping with pod scheduling decisions, allocating resources
+when requested) as well as the node level (preparing container
+startup). Such a driver can be implemented in arbitrary programming
+languages as long as it supports the resource allocation protocol and
+gRPC interfaces defined in this KEP. A utility package with Go
+support code will be made available to simplify the development of
+such a driver, but using it will not be required and its API is not
+part of this KEP.
+
+Two new API object types get added in a new API group:
+- ResourceClass, not namespaced, with the name of a resource driver and
+  parameters for the resource driver that will be used for all ResourceClaims
+  that reference this ResourceClass. These parameters might describe the kind
+  of resource that are allocated when using the ResourceClass or control
+  cluster-specific options. Because only cluster administrators are allowed to
+  create ResourceClasses and they will be passed separately to the resource
+  driver, they can be used for options that normal users should not be allowed
+  to control.
+
+  For example, a single resource driver might manage different kinds of
+  FPGAs. For each kind, one ResourceClass could define the kind and provide paths
+  to kind-specific tools or URLs for additional resources. Only a handful of
+  ResourceClasses are expected per resource driver.
+
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual metadata, a spec and a status. The
+  spec identifies the driver that handles the resource via a class
+  name.
+
+A third one with information about the driver ("ResourceDriver", similar to the
+"CSIDriver" from storage) could get added in the future. At the moment it is
+not needed yet.
+
+The ResourceClaim spec is immutable. The ResourceClaim
```

Comment:
>The ResourceClaim spec is immutable.

I get that this makes controller design for the use cases you have in mind substantially easier, and it seems reasonable to enforce (e.g. with webhooks) or state that behaviour is undefined if it is mutated. However, it seems to me that if this is adopted and shipped - with the goal of extensibility - folks will eventually want to mutate them. Analogous to resizing volumes. Random use case I can think of - if gpu supports fractional commitment, could vertically scale that commitment without restarting pod. One day ;)

Reply:
Moving parameters into a separate object would enabled resizing. We can still keep the spec immutable, but then allow resource drivers to react to changes of the parameters after allocation if they support that.

Reply:
Exactly. That's why I wanted it to be immutable as a starting point.